libgalois/src/analytics/cdlp/cdlp.cpp

```diff
@@ -0,0 +1,402 @@
+ * FITNESS FOR ANY PARTICULAR PURPOSE, NON-INFRINGEMENT AND WARRANTIES OF
+ * PERFORMANCE, AND ANY WARRANTY THAT MIGHT OTHERWISE ARISE FROM COURSE OF
+ * DEALING OR USAGE OF TRADE.  NO WARRANTY IS EITHER EXPRESS OR IMPLIED WITH
+ * RESPECT TO THE USE OF THE SOFTWARE OR DOCUMENTATION. Under no circumstances
+ * shall University be liable for incidental, special, indirect, direct or
+ * consequential damages or loss of profits, interruption of business, or
+ * related expenses which may arise from use of Software or Documentation,
+ * including but not limited to those resulting from defects in Software and/or
+ * Documentation, or loss or inaccuracy of data of any kind.
+ */
+
+#include "katana/analytics/cdlp/cdlp.h"
+
+#include <boost/unordered_map.hpp>
+
+#include "katana/ArrowRandomAccessBuilder.h"
+#include "katana/TypedPropertyGraph.h"
+
+using namespace katana::analytics;
+
+namespace {
+const unsigned int kInfinity = std::numeric_limits<unsigned int>::max();
+
+/// Limited number of iterations to limit the oscillation of the label
+/// in Synchronous algorithm. We dont need to limit it in Asynchronous algorithm.
+/// Set to 10 same as Graphalytics benchmark.
+const unsigned int kMaxIterations = 10;
+
+struct CdlpSynchronousAlgo {
+  using CommunityType = uint64_t;
+  struct NodeCommunity : public katana::PODProperty<CommunityType> {};
+
+  using NodeData = std::tuple<NodeCommunity>;
+  using EdgeData = std::tuple<>;
+  typedef katana::TypedPropertyGraph<NodeData, EdgeData> Graph;
+  typedef typename Graph::Node GNode;
+  using BiDirGraphView = katana::TypedPropertyGraphView<
+      katana::PropertyGraphViews::BiDirectional, NodeData, EdgeData>;
+
+  CdlpPlan& plan_;
+  CdlpSynchronousAlgo(CdlpPlan& plan) : plan_(plan) {}
+
+  void Initialize(Graph* graph) {
+    katana::do_all(katana::iterate(*graph), [&](const GNode& node) {
+      auto& ndata = graph->GetData<NodeCommunity>(node);
+      ndata = node;
+    });
+  }
+
+  void Deallocate(Graph*) {}
+
+  void operator()(
+      Graph* graph, const BiDirGraphView& bidir_view,
+      size_t max_iterations = kMaxIterations) {
+    if (max_iterations == 0)
+      return;
+
+    struct NodeDataPair {
+      GNode node;
+      CommunityType data;
+      NodeDataPair(GNode node, CommunityType data) : node(node), data(data) {}
+    };
+
+    size_t iterations = 0;
+    katana::InsertBag<NodeDataPair> applyBag;
+
+    /// FIXME: in this implementation, in each iteration, all the nodes are active
+    /// for gather phase. If InsertBag does not accept duplicate items then this
+    /// can be improved to have only the affected nodes to be active in next iteration
+    while (iterations < max_iterations) {
+      // Gather Phase
+      katana::do_all(
+          katana::iterate(*graph),
+          [&](const GNode& node) {
+            const auto ndata_current_comm = graph->GetData<NodeCommunity>(node);
+            typedef boost::unordered_map<CommunityType, size_t> Histogram_type;
```

Comment:
You should create a `katana::PerThreadStorage<katana::gstl::Map<K,V>> per_thread_map` and then inside the loop, you can get a thread-local instance of the map by doing ` auto& histogram = *per_thread_map.getLocal();`

Reply:
Also, use `using X = Y;` instead of `typedef Y X;`. `using` is the new preferred way of doing typedefs.